grpc-sys/src/lib.rs

```diff
@@ -379,24 +379,184 @@ pub union GrpcSliceData {
 pub enum GrpcSliceRefCount {}
 
 #[repr(C)]
-#[derive(Clone, Copy)]
-pub struct GrpcSlice {
+pub struct GrpcSliceRaw {
     ref_count: *mut GrpcSliceRefCount,
     data: GrpcSliceData,
 }
 
+/// Trivial copy. We need to copy the inlined-array, which means we can't
+/// Simply `#[derive(Clone)]`.
+/// Here we use the API provided by gRPC.
+impl Clone for GrpcSliceRaw {
+    fn clone(&self) -> Self {
+        unsafe { grpcwrap_slice_copy(self as _) }
+    }
+}
+
+impl<'a> From<&'a [i8]> for GrpcSliceRaw {
+    fn from(data: &'a [i8]) -> Self {
+        unsafe { grpc_slice_from_copied_buffer(data.as_ptr(), data.len()) }
+    }
+}
+
+impl<'a> From<&'a [u8]> for GrpcSliceRaw {
+    fn from(data: &'a [u8]) -> Self {
+        unsafe { grpc_slice_from_copied_buffer(data.as_ptr() as _, data.len()) }
+    }
+}
+
+/// A light wrapper with constructor, destructor for
+/// [GrpcSliceRaw]
+pub struct GrpcSlice {
+    raw: GrpcSliceRaw,
+}
+
+impl GrpcSlice {
+    pub fn new(len: usize) -> Self {
+        unsafe { From::from(grpc_slice_malloc(len)) }
+    }
+
+    pub fn len(&self) -> usize {
+        unsafe { grpcwrap_slice_length(&self.raw) }
+    }
+
+    pub fn as_slice(&self, offset: usize) -> &[u8] {
+        unsafe {
+            let mut len = 0;
+            let ptr = grpcwrap_slice_raw_offset(&self.raw, offset, &mut len);
+            slice::from_raw_parts(ptr as _, len)
+        }
+    }
+
+    pub unsafe fn as_ptr(&self) -> GrpcSliceRaw {
+        self.raw.clone()
+    }
+}
+
+/// Increase the ref count of the slice when cloning
+impl Clone for GrpcSlice {
+    fn clone(&self) -> Self {
+        unsafe {
```

Comment:
Looks like we can reuse the returned `GrpcSliceRaw`.

Reply:
So in order to clone `GrpcSlice`, we have to clone `GrpcSliceRaw` twice?

Reply:
`GrpcSliceRaw` is of value semantics. Clone first to inc the ref count, clone second time to create a new ref.

To reduce this (although very cheap) copy, we can make a `grpcwrap_slice_ref` which takes a pointer instead of a value.

Reply:
Re-used.